### ISSUE TITLE
`#[repr(align(8))]` is not accepted for arrays

### ISSUE THREAD

**Issue by manuthambi** [rust-lang/rust#83595]
### Code

```rust
struct Data {
    #[repr(align(8))]
    x: [usize; 20]
}
```

`#[repr(align)]` was accepted for arrays in rust v1.50, but is rejected in v1.51.

```
error[E0517]: attribute should be applied to a struct, enum, or union
 --> src/main.rs:3:12
  |
3 |     #[repr(align(8))]
  |            ^^^^^^^^
4 |     x: [usize; 20]
  |     -------------- not a struct, enum, or union
```

 I am not entirely sure whether the attribute actually did the alignment in 1.50, or the compiler just ignored it. Regardless, is this expected behavior, and if it is how do I align an array?

**Comment by ehuss** [comment]
Extra validation was added for situations where attributes were ignored (in this case it was #80641).  Previous versions issued an `unused-attribute` warning.

If you want to align an array, I think you can wrap it in a struct with `#[repr(align(8))] struct Arr([usize; 20])`.

**Comment by camelid** [comment]
Nominating to see if this was an intended "lint -> hard error" change.

(If there's another way to ask if a change is intended that people on `T-compiler` prefer, please let me know :)

@rustbot nominate

**Comment by manuthambi** [comment]
> Extra validation was added for situations where attributes were ignored (in this case it was #80641). Previous versions issued an `unused-attribute` warning.

1.50 didn't warn I think. I haven't specifically configured `unusued_attributes` in my project. Anyways, it is a good idea to lint/error for unused attributes.
 
> If you want to align an array, I think you can wrap it in a struct with `#[repr(align(8))] struct Arr([usize; 20])`.

That works, if the compiler guarantees that there will not be any padding at the beginning of `Arr` struct. Does it?

**Comment by scottmcm** [comment]
We discussed this in the lang team meeting today.

This is an intentional bugfix to resolve a place where attributes weren't being checked but were supposed to have been.  Note that this never did anything, so any `unsafe` code that was expecting this to align things is *potentially unsound*.

(There have been a few of these, recently -- for example 1.49 fixed this on enum variants.)

---

`repr(Rust)` has very few guarantees.  If you strictly need no padding at the beginning, you can use `#[repr(C, align(8))]` on the struct: https://rust.godbolt.org/z/x6648cWrE

**Comment by camelid** [comment]
Should `I-nominated` be removed now that the issue is closed?
